Move empty character cases to end

am11 opened the pull request:
Attempt to fix the issue where shortest string is being matched `vmovd` instead of `vmovdqa`. Corresponds to upstream fix: https://github.com/spadgos/regex-combiner/pull/2/commits/e083cae3bac43585c33b5cb23a637aec4112d8bb.

ashmind commented:
Thanks! Deploying it without merging is not really possible at the moment, but on the other hand manual tests are not that fun either. I would recommend a unit test: add `component.codemirror.mode-asm-instructions.tests.ts`, import the regexp, run it against e.g. `vmovdqa`, check the result -- would confirm the fix and prevent future regressions.

Btw, this might be worth trying: https://www.npmjs.com/package/regex-trie -- it seems to support character classes and `?` straight away. But no idea if it has its own issues, so another reason for auto tests.

am11 commented:
> `component.codemirror.mode-asm-instructions.tests.ts`

Going by the name of this file, I think it would be nice if we get the generated DOM output from `codemirror` in this test file, rather than manually running the raw regex. I _think_ it puts some sugar on the RegEx, like the capture grouping and `g` flag etc.

am11 commented:
Added a unittest, it fails with tip of edge branch (https://github.com/ashmind/SharpLab/commit/c8f315c85af96f9a58d35c0984ae212a518f7d1c), and fixed by the PR.

am11 commented:
> `regex-trie`

A prerequisite to use this library is that we should flatten all the nested patterns in `.txt` file, which `regex-combiner` allows but `regex-trie` does not. I have prototype ready, but it regresses a bit. I think, for now, `regex-combiner` has this main edge over `regex-trie`.

I have flattened version of .txt file (processed with https://github.com/teeshop/rexgen), but the generated regexes with both; trie and combiner, regress in size a bit.

ashmind commented on source/WebApp/tests/component.codemirror.mode-asm-instructions.tests.ts:
```diff
@@ -0,0 +1,42 @@
+
+test('mode-asm-instructions codemirror highlights correctly', async () => {
+
+    document.createRange = () => {
+        const range = new Range();
+
+        range.getBoundingClientRect = jest.fn();
+        range.getClientRects = () => ({
+            item: () => null,
+            length: 0,
+            [Symbol.iterator]: jest.fn()
+        });
+
+        return range;
+    };
```
```suggestion
    Range.prototype.getBoundingClientRect = () => fromPartial({});
    Range.prototype.getClientRects = () => fromPartial({ length: 0 });
```

ashmind commented on source/WebApp/tests/component.codemirror.mode-asm-instructions.tests.ts:
```diff
@@ -0,0 +1,42 @@
+    });
+
+    const assertHighlightedTextFor = (token: string, code: string) => {
+        const renderingListener = (instance: CodeMirror.Editor, line: CodeMirror.LineHandle, element: HTMLElement) =>
+            expect(element.innerHTML).toContain(`<span class="cm-builtin">${token}</span>`);
+
+        cm.on('renderLine', renderingListener);
+        cm.setValue(code);
+        cm.off('renderLine', renderingListener);
```
There is a small issue here -- if there is a bug where CodeMirror never calls `renderLine`, the test will still pass.
Adding a suggestion based on `jest.fn()` (that also removes some typing and need for `off`).

Could not find a good matcher alternative to the loop though.

```suggestion
        const renderingListener = jest.fn();

        cm.on('renderLine', renderingListener);
        cm.setValue(code);

        expect(renderingListener).toBeCalled();
        for (const [,, element] of renderingListener.mock.calls) {
            expect((element as HTMLElement).innerHTML).toContain(`<span class="cm-builtin">${token}</span>`);
        }
```

ashmind commented on source/WebApp/tests/component.codemirror.mode-asm-instructions.tests.ts:
```diff
@@ -0,0 +1,42 @@
+        cm.off('renderLine', renderingListener);
+    };
+
+    assertHighlightedTextFor('vmovdqa', `L0011: vmovdqa [rbp-0x30], xmm4
+        L0016: vmovdqa [rbp-0x20], xmm4
+        L001b: vmovdqa [rbp-0x10], xmm4`);
+
+    assertHighlightedTextFor('je', `L000e: je short L0015`);
+
+    assertHighlightedTextFor('jle', `L0015: jle short L0001`); 
```
These three to me are separate tests -- you can use `test.each` to run the test with each of those cases, e.g. see `server.get-branches-async.tests`. This might also simplify the assert as you can inline it.

am11 replied:
I was thinking about using something like xUnit's theory tests in jest, this is it. :)
If we want, we can inject the flat list of all ~2K instructions.

ashmind commented on source/WebApp/tests/component.codemirror.mode-asm-instructions.tests.ts:
```diff
@@ -0,0 +1,42 @@
+import CodeMirror from 'codemirror';
+import '../components/internal/codemirror/mode-asm';
+
+test('mode-asm-instructions codemirror highlights correctly', async () => {
```
Since you are testing full `mode-asm` (good choice btw!), I think the file name and test name does not have to contain `-instructions`.

```suggestion
test('mode-asm codemirror highlights correctly', async () => {
```
^ also the file name

am11 replied:
I also reworded it little explicitly what this test is trying to achieve.

ashmind commented:
> I think, for now, regex-combiner has this main edge over regex-trie.

Yep I don't mind strongly, we should look at using `?` instead of `|)` eventually, but it's out of scope for this.

am11 commented:
> `?` instead of `|)`

Just for my own understanding, how would  the `or nothing` look like in this case: `(?:ab|cd|ef|)`?

ashmind commented:
> Just for my own understanding, how would the or nothing look like in this case: `(?:ab|cd|ef|)` using `?`

It will be `(?:ab|cd|ef)?` which does not change much -- it's more for cases where `(?:)` can be dropped by `?`, e.g. `(?:x|)` => `(?:x)?` => `x?`.

am11 commented:
> It will be `(?:ab|cd|ef)?` which does not change much -- it's more for cases where `(?:)` can be dropped by `?`, e.g. `(?:x|)` => `(?:x)?` => `x?`.

Opened #592 with this optimization (shaves 260 bytes off of our regex).
